{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check how many \n",
    "\n",
    "reference_data = [\"./286.fna\"]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "for p, d in zip(val_pathogens, val_decoys):\n",
    "    print(\"Simulating data for patient\", patient_id)\n",
    "    # decide abundance of the pathogens and decoys\n",
    "    total_species_num = len(p) + len(d)\n",
    "    species_abundance = np.random.lognormal(0, 1, total_species_num)\n",
    "    species_abundance = species_abundance / np.sum(species_abundance)\n",
    "    species_abundance = np.sort(species_abundance)\n",
    "\n",
    "    # Assign pathogens the lowest abundance\n",
    "    species_index = 0\n",
    "    reads = []\n",
    "    all_species = []\n",
    "    abundance_ground_truth = {}\n",
    "    all_species.extend([(s, 'pathogen') for s in p])\n",
    "    all_species.extend([(s, 'decoy') for s in d])\n",
    "\n",
    "    # Simulate reads from pathogen genome\n",
    "    for species in all_species:\n",
    "        # genome file to sample reads from\n",
    "        if species[1] == 'pathogen':\n",
    "            species_name = pathogens_list[species[0]]\n",
    "            genome_file = os.path.join(reference_dir, \"pathogens\", species_name + \".fasta\")\n",
    "        else:\n",
    "            species_name = decoys_list[species[0]]\n",
    "            genome_file = os.path.join(reference_dir, \"commensals\", species_name + \".fasta\")\n",
    "\n",
    "        # determine number of reads of this species\n",
    "        num_reads = int(species_abundance[species_index] * READ_NUM * 4000)\n",
    "\n",
    "        # Simulate reads\n",
    "        simulated_read = subprocess.run([\"badread\", \"simulate\", \"--reference\", genome_file, \"--quantity\", str(num_reads), \"--length\", \"4000,2000\", \n",
    "                                                                \"--glitches\", \"0,0,0\", \"--junk_reads\", \"0\", \"--random_reads\", \"0\", \"--chimeras\", \"0\"], capture_output=True)\n",
    "        \n",
    "        # Store the reads in `reads` list\n",
    "        read_list = simulated_read.stdout.decode(\"utf-8\").split('\\n')\n",
    "        read_list = [r for i, r in enumerate(read_list) if i % 4 == 1]\n",
    "        reads.extend(read_list)\n",
    "\n",
    "        # store ground truth\n",
    "        abundance_ground_truth[species_name] = species_abundance[species_index]\n",
    "\n",
    "        # move on to next species\n",
    "        species_index += 1\n",
    "    \n",
    "    # Output raw reads\n",
    "    read_index = np.arange(len(reads))\n",
    "    random.shuffle(read_index)\n",
    "\n",
    "    read_i = 0\n",
    "    with open(os.path.join(output_dir, \"patient\" + str(patient_id) + \".fasta\"), 'w') as f:\n",
    "        for i, r in enumerate(read_index):\n",
    "            f.write(\">\" + str(i) + \"\\n\")\n",
    "            f.write(reads[r] + \"\\n\")\n",
    "    \n",
    "    # Output labels\n",
    "    with open(os.path.join(output_dir, \"patient\" + str(patient_id) + \"_labels.csv\"), 'w') as f:\n",
    "        f.write(\"labels\\n\")\n",
    "        for species in p:\n",
    "            f.write(pathogens_list[species] + \"\\n\")\n",
    "    \n",
    "    # Output ground truth abundance\n",
    "    with open(os.path.join(output_dir, \"patient\" + str(patient_id) + \"_abundance.csv\"), 'w') as f:\n",
    "        f.write(\"species,abundance\\n\")\n",
    "        for species in abundance_ground_truth:\n",
    "            f.write(species + \",\" + str(abundance_ground_truth[species]) + \"\\n\")\n",
    "    \n",
    "    patient_id += 1"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
